### PR TITLE
Cleanup and modernize the wrapper

### PR DESCRIPTION
**Pull request by imciner2**
Various modernizations and cleanups to the OSQP mex function and wrapper class.

* Improve readability and maintainability of the Matlab class by breaking functions into separate files
* Simplify code sections for getting OSQP constants
* Expose rho update in a nicer way
* Introduce a better and more automated wrapper for the various structs used by OSQP
* Introduce wrappers for copying/cloning vectors cleanly

### CODE REVIEW

File: @osqp/codegen.m [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+                 @(x) ischar(validatestring(x, expectedParams)));
+    addParameter(p, 'mexname', defaultMexname, @isstr);
+    addParameter(p, 'FLOAT', defaultFloat, @islogical);
+    addParameter(p, 'LONG', defaultLong, @islogical);
+    addParameter(p, 'force_rewrite', defaultFW, @islogical);
+
+    parse(p, target_dir, varargin{:});
+
+    % Set internal variables
+    if strcmp(p.Results.parameters, 'vectors')
```
Comment (AmitSolomonPrinceton) [inline @osqp/codegen.m:31]: Maybe use Matlab enumeration instead.

File: @osqp/setup.m [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        else
+            if (~isempty(A))
+                n = size(A, 2);
+            else
+                error('The problem does not have any variables');
+            end
+        end
+    else
+        n = size(P, 1);
+    end
```
Comment (AmitSolomonPrinceton) [inline @osqp/setup.m]: This might be a matter of personal taste, but this block is hard to follow because it is deeply nested. Would it be easier to read as a switch-case block instead?

File: c_sources/arrays_matlab.h
Comment (AmitSolomonPrinceton) [inline c_sources/arrays_matlab.h:1]: I don't know if it's worth the trouble, but I think having a function that checks if the input is not empty (the !in || numel==0 check) will make the code slightly more readable and easier to maintain. Please ignore this comment if you disagree.

**Review by AmitSolomonPrinceton — APPROVED** [review]
I really like the changes and the new OSQP class, it's much easier to read and cleaner than what we previously had.
The only confusing thing in my opinion is that the installation instructions in the readme file are not updated.

File: @osqp/setup.m (hunk 2 above)
Comment (imciner2) [reply @osqp/setup.m]: Yea, this block is awkward. a switch/case isn't appropriate though, since it isn't changing on values but instead based on what is empty/non-empty. I've refactored this into an if-elseif-else chain to be clearer.

File: c_sources/arrays_matlab.h
Comment (imciner2) [reply c_sources/arrays_matlab.h:1]: I think there are only two places we ever really check for this, and they are both in this file, so I don't think a helper function makes as large a difference.

File: @osqp/codegen.m (hunk 1 above)
Comment (imciner2) [reply @osqp/codegen.m:31]: This entire file needs to be rewritten to support the newer codegen paradigm anyway, so I didn't actually change anything in there. We should do this when we update that file though.

**Comment by imciner2** [comment]
>  The only confusing thing in my opinion is that the installation instructions in the readme file are not updated.

Are you referring to the ones linked to from the readme that are hosted online? We need to update those in the main OSQP repo to fix that once this is all merged.

**Comment by AmitSolomonPrinceton** [comment]
> > The only confusing thing in my opinion is that the installation instructions in the readme file are not updated.
> 
> Are you referring to the ones linked to from the readme that are hosted online? We need to update those in the main OSQP repo to fix that once this is all merged.

Both it and the instructions in the package directory.

**Comment by imciner2** [comment]
> Both it and the instructions in the package directory.

Yea, the packaging script needs a rework to do a package builder to make distribution easier.
